File: LinkedList.py
import json

class Node:
    def __init__(self, value):
        self.value = value
        self.next = None

    def json(self):
        next = self.next.value if self.next else None
        return dict(value = self.value, next = next)

    def __eq__(self, other):
            a = self.value if self else self
            b = other.value if other else other
            return a == b

class LinkedList:
    def __init__(self):
        self.head = None
        self.tail = None
        self.current = None
        self.previous = None
        self.next = None
    

    def getTail(self):
        history = []
        current = self.head
        while (current and current.next):
            history.append(current)
            for h in history:
                if(current.next == h):
                    return current
            current = current.next
        return current

    def insertNode(self, value):
        if(self.head):
            self.getTail().next = Node(value)
        else:    
            self.head = Node(value)
        return self.selectNode(self.getTail())

    def pop(self):
        current = self.head
        previous = None
        while (current and current.next):
            previous = current
            current = current.next
        if(previous):
            if(self.current == current):
                self.current = previous
            current = previous
            current.next = None
        else:
            self.head = self.current = current = None
        return current

    def selectNode(self, node):
        self.current = node
        return self.current

    def deleteList(self):
        self.head = self.current = self.previous = self.next = None
        return self

    def reverse(self):
        prev = None
        current = self.head
        while(current):
            next = current.next
            current.next = prev 
            prev = current 
            current = next
        self.head = prev
        return self
    
    def __str__(self):
        string = ""
        current = self.head
        tail = self.getTail()

        while(current != tail):
            string += str(current.value) + " -> "
            previous = current
            current = current.next
        string += str(current.value)+" -> null"  if current else "null"
        return string

    def list(self):
        history = []
        result = []
        current = self.head
        while(current and current.next):
            result.append(current.value)
            history.append(current)
            for h in history:
                if(current.next == h):
                    result.append(current.next.value)
                    return result
            current = current.next
        result.append(current.value if current else current)
        return result

    def selectHead(self):
        return self.selectNode(self.head)

    def selectNext(self):
        next = self.selectNode(self.current.next) if self.current and self.current.next else None
        return next

    def dict(self):
        result = dict()
        result['head'] = self.head
        result['current'] = self.current
        result['previous'] = self.previous
        result['next'] = self.next
        result['tail'] = self.getTail()
        return result

    def json(self):
        result = json.dumps(self.dict(), cls=ComplexEncoder, sort_keys=True, indent=4)
        return result


class ComplexEncoder(json.JSONEncoder):
    def default(self, obj):
        if hasattr(obj,'json'):
            return obj.json()
        else:
            return json.JSONEncoder.default(self, obj)


if __name__ == '__main__':
    
    ll = LinkedList()
    print(ll)
    nums = [0,1,2,3,4,5]
    for n in nums:
        ll.insertNode(n)
    print(ll)

    t = ll.getTail()
    print(ll)
    print(ll.getTail().value)

File: static/js/linkedList.js
import {redrawList, drawNode, selectNode, toggleDarkmode, pulseNode, 
    eraseNode, eraseAll, resetMatrixButtons, resetCall, 
    addToExpression, changeLabel, addToHistory, updateAttributes, drawPointer} from './animation.js'

export function sendRequest(name, callback, ...args) {
    const request = new XMLHttpRequest();
    if(args.length == 0 ){
        request.open('GET', `/${name}`);
    } else {
    request.open('GET', `/${name}?var=${args[0]}&expr=${args[1]}`);
    }
    request.onload = () => {
        const response = JSON.parse(request.responseText);
        callback(response)
    };
    request.send();
}

export function getList() {
    sendRequest('get-list', res => {
        updateAttributes(res)
        res.current && redrawList(res.current.value, setClickedCurrent)
    })
}

export function insertNode() {
    sendRequest('insert-node', res => {
        updateAttributes(res)
        if(res.current != null){
            drawNode(res.current.value)
            drawPointer(res.current.value)
            res.current.value > 0 && drawPointer(res.current.value - 1, res.current.value)
            selectNode(res.current.value)
            document.querySelector(`#Node-${res.current.value}`).onclick = () => setClickedCurrent(res.current.value)
        }
    })  
}

export function deleteList() {
    sendRequest('delete-list', res => {
        eraseAll()
        updateAttributes(res) 
        getListVars()
    })
}


//TODO can't pop self pointing node
export function popNode() {
    const getThenPop = (callback) => {
        sendRequest('get-list', res => {
            console.log(res.tail)
            res.tail && eraseNode(res.tail.value) })
            callback()
    }

    getThenPop(function() {
        sendRequest('pop-node', res => {
            updateAttributes(res) 
            res.current && selectNode(res.current.value)
        })
    }) 
    
}

export function getHead() {
    sendRequest('get-head', res => {
        const head = res.head
        updateAttributes(res) 

        if(head && head.value != null){
            selectNode(head.value)
        }
    })
};

export function getNext() {
    sendRequest('get-next', res => {
        updateAttributes(res) 
        if(res.current != null){
            selectNode(res.current.value)
        }
    })
}

export function getListVars() {
    sendRequest('get-list', res =>{
        const availableVars = [["Current", res.current], 
                            ["Previous", res.previous], 
                            ["Next", res.next],
                            ["Head", res.head]]

        availableVars.forEach((v, i) => {
            const row = i+1
            const label = v[0]
            const value = v[1] && v[1]['value'].toString()
            const btn = document.querySelector(`#var${row}-1`)

            btn.innerHTML = `${label}: ${value}`
            btn.onclick = event => {
                // First reset the other buttons
                resetMatrixButtons(event)
                //#var-{row}-{col}
                const rowCol2 = document.querySelector(`#var${row}-2`)
                rowCol2.innerHTML = label
                rowCol2.classList.remove("disabled")
                rowCol2.value = label.toLowerCase()
                rowCol2.onclick = event => addToExpression(event)
                if (value) {
                    const rowCol3 = document.querySelector(`#var${row}-3`)
                    rowCol3.innerHTML = label+'.Next'
                    rowCol3.classList.remove("disabled")
                    rowCol3.value = label.toLowerCase()+'.next'
                    rowCol3.onclick = event => addToExpression(event)
                }
                
            }
        })
    })
}

export function sendCall(statementVar, statementExpr) {
    sendRequest('modify-list', res => {
        console.log(JSON.stringify(res,null,'\t'))
        resetCall()
        getListVars(res)
        addToHistory(statementVar, statementExpr)
        updateAttributes(res) 
        Object.keys(res.diff).forEach( diff => {
            switch (diff) {
                case 'current':
                    console.log(res.current)
                    selectNode(res.current ? res.current.value : res.current)
                    break;
                case 'previous':
                    res.previous && changeLabel(res.previous.value,"P")
                    break;
                case 'next':
                    res.next && changeLabel(res.next.value,"N")
                    break;
                case 'head':
                    res.head && changeLabel(res.head.value,"H")
                    break;
                case diff.split('.')[0]+'.next':
                    const newNext = res[diff.split('.')[0]].next//res.current.next
                    const nodeValToChange = res[diff.split('.')[0]].value
                    console.log(`Change an arrow pointer now for  ${diff}`)
                    console.log(`Point ${nodeValToChange} to ${newNext}`)
                    drawPointer(nodeValToChange,newNext)
                    break;
                default:
                    // Nothing to change
                    break;
            }
        })
    },
    statementVar, statementExpr)
}

export function setClickedCurrent(val) {
    const statementVar = "current"
    const statementExpr = val
    sendCall(statementVar, statementExpr) 
}
